docs/proposed/psa-driver-interface.md

```diff
@@ -536,6 +537,37 @@ The core will only call `acme_export_public_key` on a private key. Drivers imple
 
 The core guarantees that the size of the output buffer (`data_size`) is sufficient to export any key with the given attributes. The driver must set `*data_length` to the exact size of the exported key.
 
+### Builtin keys for opaque drivers
+
+An opaque driver can be pre-provisioned from the factory with keys. These keys are never seen by the PSA Crypto core, but in order to be able to use these keys, they need to be declared to the core.
+
+An opaque driver can declare such keys in its JSON descriptor file. The declaration is an array of strings, where each string is the name of a preprovisioned key, e.g.:
+```
+{
+    "builtin_keys": ["pre_key_1", "pre_key_2"]
+}
+```
+
+The key names need to be compatible with C language variable naming constraints, i.e. they cannot contain spaces and cannot start with a digit.
+
+For each declared key, the respective driver needs to define and externally declare the corresponding PSA Crypto key attributes (`psa_key_attributes_t`). The name of this variable, for each key, is `{prefix}_builtin_{keyname}_attributes`, where `{prefix}` is the implementing driver's prefix (lowercase), and `{keyname}` is the lowercase name of the key from the `builtin_keys` array.
+
+E.g. for a key called `"identity"`, a driver called `"acme"` would need to expose:
+```
+extern psa_core_key_attributes_t acme_builtin_identity_attributes;
+
+psa_key_attributes_t acme_builtin_identity_attributes = {
+    ...
+};
+```
+
+In order to use these keys, the JSON conversion process will generate a preprocessor symbol for each key, called `{PREFIX}_BUILTIN_{KEYNAME}_ID`, where `{PREFIX}` is the uppercase driver prefix, and `{KEYNAME}` the uppercase name of the key. This symbol yields a `psa_key_id_t` from the PSA key ID vendor range, and can be supplied as an argument to `psa_open_key()` to get a handle to access the builtin key.
```

Comment:
How would the key id be assigned? Is `{PREFIX}_BUILTIN_{KEYNAME}_ID` guaranteed to be a compile-time constant? If it is (and I think it makes sense for it to be), I don't know how to ensure that key ids are stable on a rebuild in practice. It's easy to ensure that key ids are stable on a rebuild with an identical configuration, but not so when there's a configuration change, for example because the driver has been updated and now declares an extra key.

Reply:
Do they need to be stable on regeneration of the driver set? Regeneration will touch the header file where the IDs are declared, which should (when 'making' your application) result in your application being recompiled with the new IDs.

Since the IDs aren't stored non-volatile, I don't think this is a problem?

Reply:
I'd expect to be able to rebuild the application and the library+driver separately. This is especially important when the crypto is in a PSA service, for example running in a secure enclave.

Reply:
You already cannot do that when driver lifetimes are auto-generated by the JSON.

Or are you referring to wanting the same escape hatch applied here as with the driver lifetimes, i.e. optionally being able to specify a specific key ID for each builtin key? That'll get messy too with trying to coordinate usage of the vendor key ID space. Which brings us full circle to my initial proposal of carving up that space and building the key ID using the driver lifetime value and a key 'sequence number'. Then it's just a matter of specifying new keys need to be added additively on a driver update.

Reply:
The driver description is defined as an ordered list. An implementation can choose to assign locations based on the order of in this list, and the locations will be stable across rebuilds as long as the build process includes the same drivers in the same order. This is a fairly low bar to meet. [I think this should be made a requirement](https://github.com/ARMmbed/mbedtls/issues/3775).

I don't see how to get the same level of stability for automatic key id assignments. It's feasible to mandate that the exact same driver description list will give consistent assignments, but plausible changes are problematic, e.g. driver 1 has a new key and this shifts the key ids of driver 2.

[Namespacing key ids by location as you proposed](https://github.com/ARMmbed/mbedtls/issues/3285#issuecomment-700677456) solves this but it's a tight fit with 20-bit vendor ids and 24-bit locations.

Reply:
So what's your suggestion then, seeing you're not a fan of either? Statically define key IDs in the driver JSON? How will those be conflict-free across vendors?